### DISCUSSION TITLE
How to make Tailwind CSS apply to remark-directive elements?

### DISCUSSION THREAD

**Discussion by edward1127** [remarkjs/remark#1171] [Q&A] [answered]
I use remark-directive to edit markdown content. When I use syntax like `::::div{.bg-primary}`, the style is only applied if it is already loaded with my custom component. How can I preload all styles similar to what Tailwind and Typography achieve?

```
import "katex/dist/katex.min.css";
import React from "react";
import remarkMath from "remark-math";
import remarkGfm from "remark-gfm";
import rehypeKatex from "rehype-katex";
import ReactMarkdown from "react-markdown";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import NumericInput from "@/components/Control/NumericInput";
import { IQuestion } from "@/interfaces/problems";
import TextInput from "@/components/Control/TextInput";
import CheckBox from "@/components/Control/CheckBox";
import TextArea from "./Control/TextArea";

export default function QuestionRender({ question }: { question: IQuestion }) {
  const renderParagraph = (props) => <div>{props.children}</div>;
  const components = {
    "numeric-input": (props) => <NumericInput question={question} {...props} />,
    "text-area": (props) => <TextArea question={question} {...props} />,
    "text-input": (props) => <TextInput question={question} {...props} />,
    checkbox: (props) => <CheckBox question={question} {...props} />,
    p: renderParagraph,
  };
  return (
    <ReactMarkdown
      remarkPlugins={[
        remarkDirective,
        remarkDirectiveRehype,
        remarkMath,
        remarkGfm,
      ]}
      rehypePlugins={[rehypeKatex]}
      components={components}
    >
      {question.description}
    </ReactMarkdown>
  );
}

```

**Comment by Murderlon** [comment]
If I understand correctly, this is more a Tailwind question than a unified one. You successfully applied the class to your directive HTML output but Tailwind didn't output the class? If that's the case, you probably need to configure your `content` in `tailwind.config.js` to also run against markdown files (or the HTML output).

**Reply by edward1127** [reply]
Sorry, let me clarify my question.
![image](https://github.com/remarkjs/remark/assets/45161984/fc36646f-94cb-49ca-8c23-fe2992a91c6c)
![image](https://github.com/remarkjs/remark/assets/45161984/438b4c09-2f52-4b00-b670-c74203f694a7)
I guess because the button uses bg-primary so the style is loaded already, and the style can be applied. However, since bg-red-200 is not loaded anywhere before so even if the className is shown there it cannot find the corresponding style.

**Reply by Murderlon** [reply]
This seems to have nothing to do with the unified ecosystem. I would recommend making sure your development server recompiles tailwind on save and that these classes are sourced in the needed files in your tailwind config.

**Reply by edward1127** [reply]
thanks!

**Comment by remcohaszing** [comment, marked as answer]
Tailwind scans your source code at build time and generates the CSS matching. Your example uses `react-markdown` to process a markdown document on the client side. Its contents are never scanned by Tailwind.

If you use `react-markdown` to render markdown files from your repository, I recommend switching to [MDX](https://mdxjs.com). You don’t need to use MDX files if you don’t want to. It can load plain markdown files as React components too.

If you need to render dynamic content, you need to run Tailwind in the browser. This is not officially supported. You can draw some inspiration from [`monaco-tailwindcss`](https://github.com/remcohaszing/monaco-tailwindcss). Most notably you should have a look at [`generateStylesFromContent`](https://github.com/remcohaszing/monaco-tailwindcss/blob/main/src/tailwindcss.worker.ts#L199-L210) and the various hacks specified in [`build.js`](https://github.com/remcohaszing/monaco-tailwindcss/blob/main/build.js). Alternatively you could use [`jit-browser-tailwindcss`](https://github.com/mhsdesign/jit-browser-tailwindcss).

**Reply by edward1127** [reply]
thanks for pointing out the issue and the direction.

**Reply by edward1127** [reply]
I end up make it work with the following code

```
"use client";
import "katex/dist/katex.min.css";
import {
  TailwindConfig,
  createTailwindcss,
} from "@mhsdesign/jit-browser-tailwindcss";
import React, { useEffect, useState } from "react";
import remarkMath from "remark-math";
import remarkGfm from "remark-gfm";
import rehypeKatex from "rehype-katex";
import ReactMarkdown from "react-markdown";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import NumericInput from "@/components/Control/NumericInput";
import { IQuestion } from "@/interfaces/problems";
import TextInput from "@/components/Control/TextInput";
import CheckBox from "@/components/Control/CheckBox";
import TextArea from "./Control/TextArea";

export default function QuestionRender({ question }: { question: IQuestion }) {
  const renderParagraph = (props) => <div>{props.children}</div>;
  const components = {
    "numeric-input": (props) => <NumericInput question={question} {...props} />,
    "text-area": (props) => <TextArea question={question} {...props} />,
    "text-input": (props) => <TextInput question={question} {...props} />,
    checkbox: (props) => <CheckBox question={question} {...props} />,
    p: renderParagraph,
  };

  const [style, setStyle] = useState("");

  useEffect(() => {
    const createStyleFromTailwind = async () => {
      const tailwindConfig: TailwindConfig = {
        theme: {},
      };
      const tailwindCss = createTailwindcss({ tailwindConfig });

      const contentElements = document.querySelectorAll(".dynamic-tailwind *");

      const content = Array.from(contentElements).reduce(
        (carry, el) => carry + el.outerHTML,
        ""
      );
      const css = await tailwindCss.generateStylesFromContent(
        `
      @tailwind components;
      @tailwind utilities;
      `,
        [content]
      );
      if (css === "\n      ") {
        return;
      }
      setStyle(css);
    };

    createStyleFromTailwind();
  }, [question.description]);

  return (
    <div>
      <ReactMarkdown
        className={`prose max-w-none dynamic-tailwind text-black`}
        remarkPlugins={[
          remarkDirective,
          remarkDirectiveRehype,
          remarkMath,
          remarkGfm,
        ]}
        rehypePlugins={[rehypeKatex]}
        components={components}
      >
        {question.description}
      </ReactMarkdown>
      <style jsx>{`
        ${style}
      `}</style>
    </div>
  );
}

```
